### PR TITLE
Refactoring of the `pbkdf2` crate.

### PR DESCRIPTION
**Pull request by jfrimmel**
This PR applies refactoring to the `pbkdf2` crate, especially for the `simple` module.
It is best viewed commit by commit.

### CODE REVIEW

**Comment by jfrimmel** [comment]
Do you want to keep the minimum Rust version of 1.22? Some dependencies (e.g. rand) could also be updated, if the MSRV is bumped

File: pbkdf2/src/simple.rs [hunk 1]
```diff
@@ -69,65 +70,26 @@ pub fn pbkdf2_simple(password: &str, c: u32) -> io::Result<String> {
...
-            Ok(salt) => salt,
-            Err(_) => return Err(CheckError::InvalidFormat)
-        },
-        None => return Err(CheckError::InvalidFormat)
+pub fn pbkdf2_check(password: &str, hashed_value: &str) -> Result<(), CheckError> {
+    let parts: Vec<_> = hashed_value.split('$').collect();
+
+    // check the format of the input: there may be no tokens before the first
+    // and after the last `$`, tokens must have correct information and length.
+    let (count, salt, hash) = match parts[..] {
```
Comment (newpavlov) [inline pbkdf2/src/simple.rs]: I think you can remove the allocation by using:
```rust
let mut p = hashed_value.split('$');
let buf: [Option<&str>; 8] = [
    p.next(), p.next(), p.next(), p.next(),
    p.next(), p.next(), p.next(), p.next(),
];
let (count, salt, hash) = match buf {
    [
        Some(""), Some("rpbkdf2"), Some("0"), Some(count),
        Some(salt), Some(hash), Some(""), None,
    ] => (count, salt, hash),
    _ => return Err(CheckError::InvalidFormat),
};
```

Also it looks like `Err(CheckError::InvalidFormat)?` is discouraged (e.g. by clippy), so we probably better to replace it with an explicit `return`.

Comment (jfrimmel) [reply pbkdf2/src/simple.rs]: Thank you for your suggestion. I will remove the unnecessary allocation (I wasn't happy with it, too) by applying your changes :smile: 
Unfortunately, I don't think, that Rust 1.22 will support this code as well.

**Comment by jfrimmel** [comment]
> This will fix #17, correct?

Unfortunately no, since this PR does not increase the MSRV. This would be required to use Rand 0.7.

**Comment by tarcieri** [comment]
@jfrimmel can you rebase? we've bumped MSRV to 1.41+

**Comment by newpavlov** [comment]
I've included your refactor into #33.

**Pull request closed without merging**